Added the schema file and how the features will be implemented using …

GauriSaran opened the pull request:
Added shared_data.csv file which contains the schema and the predefined rules for sharing. The Readme document is modified to add the details on how the features defined in schema_features branch will be implemented using CSV schema and the web application tool that data custodian can use. The schema has been explained with all the rules.

DougManuel commented on sharing/Readme_sharing.md:
```diff
@@ -1,31 +1,145 @@
+
+**is_excluded:** This rule will allow the user to exclude any variable from their data that they do not wish to share with any data repository or organizations at all. The rule will accept values as True/ False based on whether the variable can be shared or not shared.
+
+**shared_with_Public:** This rule will allow the user to exclude any variable that the data custodian may not wish to share with the general public based on True/ False values specified for that rule in the sharing schema.
```
I suggest that we retire the historic sharing approach in WWMeasure. I am not aware of any user who is using those sharing rules. @jeandavidt  what do you think?

DougManuel replied:
I suggest that we are flexible and allow on sharing schema to be used for multiple end-users. We would use your approach of the prefix `share_with_...` to identify an end-user. Who the data is shared with is defined by the custodian. Meaning, any header with `share_with_` identifies that header as a sharing entity. 

i.e. `share_with_myFriend-A`, `share_with_myFriend-B`, 

A data custodian could also have multiple schemas if they wish. They could, for example, have a separate schema, `share_data_myFriend-A.csv`and `share_data_myFriend-B.csv`.

GauriSaran replied:
Yes that will work, because the way code is structured, it will not be hard coded with whom they share. But what is essential is that the last part of the name of who they share with should be what they input into the web application form. This way code will figure out who they are sharing with as a different access group.

DougManuel replied:
Hmm. We may want to include a `sharing_with...` label and/or description with the schema. The label and description would be used with the web app and other uses.

`sharing_with_Friend-A`, `label_Friend-A`, `description_Friend-A`.

GauriSaran replied:
should I write it as only one sharing rule in the specification document such as **sharing_with_label** where label is who they are sharing the data with. This will cut down on the number of sharing rules we have in specification document because the user will be able to make as many labels as they want when they work with the schema, so I can simply add it as a single generalized property or rule: "sharing_with_label** in the specification?

GauriSaran commented:
@DougManuel This branch is totally changed now as we updated schema style from wide to long table as per @yulric suggestions. I have rewritten the entire specification on how features will be implemented. Please take a look and let me know if it sounds fine. Thank you.

GauriSaran commented:
I do have one question. If the data can be entered in either excel or csv format, then if the user needs to enter several different tables to filter data for, then will they be input a separate file for each table on the web application tool? For excel I believe they can use one workbook with different sheets, but for csv file they will need to input separate files for separate tables I assume?

yulric reviewed (changes requested):
1. There should be a new section after the "Features" section called "Sharing CSV Schema". Its sub sections and what goes in them are given below:
    1. **Introduction**: About one paragraph. Describes the purpose of the file and what each row represents. Finally, outlines the three step process for creating a rule we went over
    2. The next three sections should describe each step of creating the rule, going over the columns needed for each step. When making these sections try to break it down into the columns part of each section and the different features they can accomplish with these columns. Also provide examples for each feature like you did in this document.
    3. A section that goes over any extra columns not covered in the last three sections
    4. A final section with the list of columns and their descriptions
2. We do not need the "How the features will be implemented?" section, it will come in the next milestone. You can remove it
3. We do not need the "Context" section, we can add that later after the "Aim/Objective" section. 
    1. You can remove the first three points, 9th and 16th point
    2. All the other points should into one of the three sections describing the steps to make rule 
4. The **Sharing processes** section will be made part of the section where we go over the different rules
5. The **Sharing schema and rules** section can be converted to the final section

GauriSaran commented:
@yulric  Made the changes. Please take a look. Thanks

jeandavidt commented:
> I suggest that we retire the historic sharing approach in WWMeasure. I am not aware of any user who is using those sharing rules. @jeandavidt  what do you think?
 

I think that removing the old sharing approach altogether is indeed a good idea 👍

yulric commented:
@GauriSaran I've made changes to this section. Please pull my commit and read through it. Let me know if there are any issues. I've added a new column called **direction** and also changes the column names to be camelCase to be consistent with the ODM.

@DougManuel This PR should be ready for you to review again.

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+| 1      | University of Laval  | Laval Public Health    | school |
+
+
+1. Remove rows whose site ID in the Sample table is 2 for Ottawa Public Health (OPH)
```
`site ID`  --> `siteID`

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+* In examples 2 and 4 where multiple columns and tables were selected respectively, a **;** was used to seperate the values. The same symbol was used to seperate multiple organizations in the previous step. In fact, throughout the entire document when multiple values need to put into a cell, the **;** symbol should be used to seperate them.
+* In examples 3 and 5 where all the columns in a table and all the tables were selected respectively, the keyword **all** was used. Similar to the **;** symbol, the keyword **all** may be used in a cell to mean everything.
+
+### 3. Selecting the Direction of the Rule
```
I like this concept but I don't see how it is implemented below. 

Generally, in data science we think of:
`filter` --> column 
`select` --> row. 

What about changing this terms from `direction` to `filter and select`?

DougManuel replied:
I'm going to backtrack on my suggestion. The 'direction' of the rule could work well. It is simple and inuitive.

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+| siteID | name                 | publicHealthDepartment | type   |
+|--------|----------------------|------------------------|--------|
+| 1      | University of Ottawa | Ottawa Public Health   | school |
+| 1      | University of Laval  | Laval Public Health    | school |
```
Why do University of Ottawa and University of Laval have the same siteID? 

Shouldn't they be unique? i.e. `1` and `2`?

Did you intend Univeristy of Ottawa `siteID` = 2?

yulric replied:
Correct, I'll fix that.

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+
+    | sharedWith | tableName | variableName | direction | filterValue |
+    |------------|-----------|--------------|-----------|-------------|
+    | OPH        | Sample    | siteID       | row       | 1           |
```
This is a situation where it may be helpful to have a 'not' statement (or filter in). If you want to share only OPH data with OPH and their siteID = 2, then an efficient way to share is `filterValue` = ![2] or not[2]. Filter all data that is not siteID = 2. 

What do you think?

yulric replied:
Makes sense, I'll add that in.

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+    |------------|-----------|--------------|-----------|--------------|
+    | LPH        | all       | type         | column    | all          |
+
+4. Remove samples taken in the year 2021 and those whose volume is less than 5 for OPH and PHAC
```
#4 will require more discussion. We will need a process of linking the two sharing rules together (the two rows in the sharing table). Also what if this is an 'or' filter?

DougManuel replied:
Here is an example that would be ideal to accommodate:

For Ottawa, exclude rows and columns in WWMeasure  in March 2021 ANDS have [`qualityFlag` = False  OR LOD < 100]. Exclude all other rows and columns in other tables during the same period with uuID that were excluded in the above rule.

This may be too tricky a rule to include, but it is a real example from Ottawa during the spring where there the spring melt. The Ottawa lab may not what to include the results from that melt period where the testing quality was poor.

DougManuel replied:
The validation approach in the Cerebrus package can accommodate this type of complex rule. Right?

Another typical approach that users are familiar with is numbering each row in share.csv and then create a rule with AND and OR. 

rule1 AND [rule2 or rule3].

As a general rule, we may want to add a unique ID to the rule `ruleID`. This allows people to more easily reference, duplicate, update, share rules. The ruleID would then facilitate more complicated sharing schema.

yulric replied:
This was a mistake, the example should be "Remove samples taken in the year 2021 **or** those whose volume is less than 5 for OPH and PHAC". Two different rules are implicitly combined using the OR operator and not AND.

With regards to your other points, I agree, we need to think of a solution "combine" rules together if we want to accommodate all scenarios. I was thinking along the same lines as your previous point but I'll address in a new PR since it would be an addon.

I'll add in a **ruleId** column.

DougManuel commented on sharing/README.md:
```diff
@@ -25,3 +25,172 @@ High level features include:
+    | LPH        | Site      | notes                  | column    | all                     |
+    | LPH        | Sample    | publicHealthDepartment | row       | Ottawa Public Health    |
+
+## Sharing CSV Columns
```
The names in this section are different from the `sharing_data.csv` file.

Maybe have this section earlier?

yulric replied:
The `shared_data.csv` file is old. I'll delete it.

DougManuel reviewed (changes requested):
A) The overall structure for `shared_data.csv` looks good. Multiple shared agencies in each row works well. Could we call it just `sharing.csv`?

B) A few minor wording changes:
1)  headers should be pascal case. i.e. `shared_with` should be `sharedWith`.
2) suggest shortening some names:
     `tableName` --> `table`
     `variableName` —> `variable`
     `filter_values` —> `escludeRule`
     `description_licence` —> `license`
3) `shared_with` —> `shareWith`

C) `shareWith`:   Somehow we will need more information about who the data is being shared with: contact information, organization full name, and description, etc. This feels like another look-up table -- not sure. We could address this issue in a future update.

yulric commented:
> A) The overall structure for `shared_data.csv` looks good. Multiple shared agencies in each row works well. Could we call it just `sharing.csv`?
> 
> B) A few minor wording changes:
> 
> 1. headers should be pascal case. i.e. `shared_with` should be `sharedWith`.
> 2. suggest shortening some names:
>    `tableName` --> `table`
>    `variableName` —> `variable`
>    `filter_values` —> `excludeRule`
>    `description_licence` —> `license`
> 3. `shared_with` —> `shareWith`
> 
> C) `shareWith`: Somehow we will need more information about who the data is being shared with: contact information, organization full name, and description, etc. This feels like another look-up table -- not sure. We could address this issue in a future update.

Agreed with everything except renaming **filterValues** to **excludeRule**. If we're going with the terminology that a rule is a row in the file, then **excludeRule** does not fit with the usage of this column. Here, the user adds in all the values for the selected entities that they want to be part of the rule. We could rename it to **ruleValues**?

yulric commented:
@GauriSaran I've pushed commits to address Doug's comments. Look at the message in https://github.com/Big-Life-Lab/ODM/pull/159/commits/ccd2f6fd5e66f4efbc910e206e63303a8f4c1021 for those that may change your implementation. Can you update the sharing_functions branch with these updates? I believe the biggest one will be the addition of the NOT operator (Look at lines 130 to 136 for an example). Let me know when you're done and I'll review it.

yulric commented:
> @DougManuel and @yulric I think we should change the name of `sharedWith` property too. Because to me it feels like inclusion not exclusion rule even though I am applied an exclusion rule in practice. Any suggestions on that?

We'll keep it with `sharedWith` for now. This column is about which organizations this rule applies rather than anything about exclusions.

I've also made the ruleValue changes and will merge this PR in.
